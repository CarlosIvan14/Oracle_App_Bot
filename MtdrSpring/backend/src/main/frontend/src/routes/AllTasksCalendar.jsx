import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";

export default function AllTasksCalendar() {
  const { sprintId } = useParams();
  const [assignedTasks, setAssignedTasks] = useState([]);
  const [unassignedTasks, setUnassignedTasks] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");
  const [selectedTask, setSelectedTask] = useState(null);

  useEffect(() => {
    const fetchTasks = async () => {
      try {
        setLoading(true);

        const [assignedResponse, unassignedResponse] = await Promise.all([
          fetch(`http://140.84.170.68/api/task-assignees/by-sprint/${sprintId}`),
          fetch(`http://140.84.170.68/api/tasks/unassigned/${sprintId}`)
        ]);

        if (!assignedResponse.ok || !unassignedResponse.ok) {
          throw new Error("Error al cargar las tareas");
        }

        const assignedData = await assignedResponse.json();
        const unassignedData = await unassignedResponse.json();

        const transformedAssigned = assignedData.map((item) => ({
          ...item.task,
          assignee: item.projectUser.user.name,
          assigneeDetails: item.projectUser.user,
        }));

        setAssignedTasks(transformedAssigned);
        setUnassignedTasks(unassignedData);
      } catch (e) {
        setError(e.message);
      } finally {
        setLoading(false);
      }
    };

    fetchTasks();
  }, [sprintId]);

  if (loading)
    return <p className="text-center mt-8 text-white">Cargando tareas…</p>;
  if (error) return <p className="text-center mt-8 text-red-500">{error}</p>;

  const renderTaskCard = (task) => (
    <div
      key={task.id}
      className="bg-black bg-opacity-20 p-4 rounded-2xl flex flex-col space-y-2"
    >
      <h3 className="font-semibold text-white truncate">{task.name}</h3>
      <p className="text-gray-300 truncate">
        Desarrollador: {task.assignee || "Libre"}
      </p>
      <p className="text-gray-400 text-sm">
        Story Points:{" "}
        <span
          className={`inline-block px-2 py-0.5 rounded-full text-white ${
            task.storyPoints <= 3
              ? "bg-green-500"
              : task.storyPoints <= 6
                ? "bg-yellow-500"
                : "bg-red-500"
          }`}
        >
          {task.storyPoints}
        </span>
      </p>
      <p className="text-gray-400 text-sm">Estimadas: {task.estimatedHours}h</p>
      <p className="text-gray-400 text-sm">Reales: {task.realHours || 0}h</p>
      <button
        className="mt-auto self-end text-blue-300 underline"
        onClick={() => setSelectedTask(task)}
      >
        Ver más
      </button>
    </div>
  );

  return (
    <div className="p-6">
      <h1 className="text-3xl font-bold mb-4 text-white">
        Todas las tareas del Sprint {sprintId}
      </h1>

      {/* ==== CONTENEDOR ÚNICO ==== */}
      <h2 className="text-xl font-semibold mb-2 text-white">Tareas</h2>
      <div
        className="no-white-scrollbar overflow-y-auto max-h-[calc(100vh-220px)] pr-2"
        style={{
          scrollbarWidth: "thin",
          scrollbarColor: "rgba(107,114,128,0.6) transparent",
        }}
      >
        {/* Asignadas primero */}
        {assignedTasks.length > 0 && (
          <div className="mb-6">
            <h3 className="text-lg font-semibold mb-2 text-white">
              Tareas Asignadas
            </h3>
            <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6">
              {assignedTasks.map(renderTaskCard)}
            </div>
          </div>
        )}

        {/* Libres al final */}
        {unassignedTasks.length > 0 && (
          <div className="mt-6">
            <h3 className="text-lg font-semibold mb-2 text-white">
              Tareas Libres
            </h3>
            <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6">
              {unassignedTasks.map(renderTaskCard)}
            </div>
          </div>
        )}
      </div>

      {/* ==== MODAL ==== */}
      {selectedTask && (
        <div
          className="fixed inset-0 z-50 flex items-center justify-center bg-black bg-opacity-70"
          onClick={() => setSelectedTask(null)}
        >
          <div
            className="max-w-md w-full p-6 bg-white dark:bg-customDark dark:text-white rounded-2xl"
            onClick={(e) => e.stopPropagation()}
          >
            <h2 className="text-2xl font-bold mb-4">{selectedTask.name}</h2>
            <p className="mb-2">
              <span className="font-semibold">Status:</span>{" "}
              {selectedTask.status}
            </p>
            <p className="mb-2">
              <span className="font-semibold">Desarrollador:</span>{" "}
              {selectedTask.assignee}
            </p>

            {selectedTask.assigneeDetails && (
              <>
                <p className="mb-2">
                  <span className="font-semibold">Email:</span>{" "}
                  {selectedTask.assigneeDetails.email}
                </p>
                <p className="mb-2">
                  <span className="font-semibold">Rol:</span>{" "}
                  {selectedTask.assigneeDetails.role}
                </p>
              </>
            )}

            <p className="mb-2">
              <span className="font-semibold">Story Points:</span>{" "}
              <span
                className={`inline-block px-2 py-0.5 rounded-full text-white ${
                  selectedTask.storyPoints <= 3
                    ? "bg-green-500"
                    : selectedTask.storyPoints <= 6
                      ? "bg-yellow-500"
                      : "bg-red-500"
                }`}
              >
                {selectedTask.storyPoints}
              </span>
            </p>
            <p className="mb-2">
              <span className="font-semibold">Horas Estimadas:</span>{" "}
              {selectedTask.estimatedHours}h
            </p>
            <p className="mb-2">
              <span className="font-semibold">Horas Reales:</span>{" "}
              {selectedTask.realHours || 0}h
            </p>
            <p className="mb-4">
              <span className="font-semibold">Deadline:</span>{" "}
              {new Date(selectedTask.deadline).toLocaleString()}
            </p>
            {selectedTask.description && (
              <p className="mb-4">
                <span className="font-semibold">Descripción:</span>{" "}
                {selectedTask.description}
              </p>
            )}
            <button
              onClick={() => setSelectedTask(null)}
              className="mt-4 rounded-full border border-gray-300 px-4 py-2 hover:bg-gray-200 dark:hover:bg-customDarkligth"
            >
              Cerrar
            </button>
          </div>
        </div>
      )}
    </div>
  );
}
