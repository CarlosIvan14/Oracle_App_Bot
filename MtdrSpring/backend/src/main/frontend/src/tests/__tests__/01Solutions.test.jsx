//01Solutions.test.jsx
// 1. Visualización en tiempo real de tareas asignadas a cada usuario.

/*
    Involved files:
        - src/App.jsx
        - src/routes/AllTasksCalendar.jsx
*/

import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import "@testing-library/jest-dom";

import AllTasksCalendar from "../../routes/AllTasksCalendar";
import { server } from "../testServer";

/**
 * Inicia el servidor mock antes de todos los tests y lo cierra al finalizar.
 */
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

/**
 * Renderiza el componente en un contexto de router,
 * posicionándose en la ruta especificada para simular acceso por sprintId.
 *
 * @param ui Componente React a renderizar.
 * @param options.route Ruta inicial (por defecto "/calendar/123").
 * @returns {RenderResult} Resultado de renderizado de Testing Library.
 */
const renderWithRouter = (ui, { route = "/calendar/123" } = {}) => {
  window.history.pushState({}, "Test page", route);
  return render(
    <MemoryRouter initialEntries={[route]}>
      <Routes>
        <Route path="/calendar/:sprintId" element={ui} />
      </Routes>
    </MemoryRouter>
  );
};

describe("Componente AllTasksCalendar", () => {
  beforeAll(() => {
    // Silencia warnings irrelevantes de React Router
    jest.spyOn(console, "warn").mockImplementation((msg) => {
      if (msg.includes("React Router Future Flag Warning")) return;
      console.warn(msg);
    });
  });

  test("muestra tareas asignadas y sin asignar correctamente y coincide con el snapshot", async () => {
    // Desestructuramos asFragment para snapshot
    const { asFragment } = renderWithRouter(<AllTasksCalendar />);

    // Verifica el estado de carga inicial
    expect(screen.getByText(/cargando tareas/i)).toBeInTheDocument();

    // Espera a que termine la carga
    await waitFor(() =>
      expect(screen.queryByText(/cargando tareas/i)).not.toBeInTheDocument()
    );

    // Captura el snapshot de la UI cargada
    expect(asFragment()).toMatchSnapshot();

    // Confirma tarea con desarrollador asignado
    expect(screen.getByText("Implement login")).toBeInTheDocument();
    expect(screen.getByText(/Desarrollador: Alice/i)).toBeInTheDocument();

    // Confirma tarea sin desarrollador (Libre)
    expect(screen.getByText("Fix logout bug")).toBeInTheDocument();
    expect(screen.getByText(/Desarrollador: Libre/i)).toBeInTheDocument();
  });
});
