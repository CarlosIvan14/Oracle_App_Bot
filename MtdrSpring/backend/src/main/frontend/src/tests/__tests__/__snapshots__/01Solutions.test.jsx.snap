// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente AllTasksCalendar muestra tareas asignadas y sin asignar correctamente y coincide con el snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-6"
  >
    <h1
      class="text-3xl font-bold mb-4 text-white"
    >
      Todas las tareas del Sprint 123
    </h1>
    <h2
      class="text-xl font-semibold mb-2 text-white"
    >
      Tareas
    </h2>
    <div
      class="no-white-scrollbar overflow-y-auto max-h-[calc(100vh-220px)] pr-2"
    >
      <div
        class="mb-6"
      >
        <h3
          class="text-lg font-semibold mb-2 text-white"
        >
          Tareas Asignadas
        </h3>
        <div
          class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6"
        >
          <div
            class="bg-black bg-opacity-20 p-4 rounded-2xl flex flex-col space-y-2"
          >
            <h3
              class="font-semibold text-white truncate"
            >
              Test Task
            </h3>
            <p
              class="text-gray-300 truncate"
            >
              Desarrollador: Test User
            </p>
            <p
              class="text-gray-400 text-sm"
            >
              Story Points: 
              <span
                class="inline-block px-2 py-0.5 rounded-full text-white bg-green-500"
              >
                3
              </span>
            </p>
            <p
              class="text-gray-400 text-sm"
            >
              Estimadas: 4h
            </p>
            <p
              class="text-gray-400 text-sm"
            >
              Reales: 0h
            </p>
            <button
              class="mt-auto self-end text-blue-300 underline"
            >
              Ver más
            </button>
          </div>
        </div>
      </div>
      <div
        class="mt-6"
      >
        <h3
          class="text-lg font-semibold mb-2 text-white"
        >
          Tareas Libres
        </h3>
        <div
          class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6"
        >
          <div
            class="bg-black bg-opacity-20 p-4 rounded-2xl flex flex-col space-y-2"
          >
            <h3
              class="font-semibold text-white truncate"
            >
              Fix logout bug
            </h3>
            <p
              class="text-gray-300 truncate"
            >
              Desarrollador: Libre
            </p>
            <p
              class="text-gray-400 text-sm"
            >
              Story Points: 
              <span
                class="inline-block px-2 py-0.5 rounded-full text-white bg-green-500"
              >
                2
              </span>
            </p>
            <p
              class="text-gray-400 text-sm"
            >
              Estimadas: 3h
            </p>
            <p
              class="text-gray-400 text-sm"
            >
              Reales: 0h
            </p>
            <button
              class="mt-auto self-end text-blue-300 underline"
            >
              Ver más
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
